Radiation API

DustyDave961 opened the pull request:
Radiation resistance API made with the help of Microsoft Copilot to fix #605.

DustyDave961 committed:
Update radiation.lua

Api added with the help of Copilot AI.

DustyDave961 committed:
Update nodes.lua

"rad_resistance" group added to lead block node definitions. Playtesting required.

DustyDave961 committed:
Update radiation.lua

Lead block removed from node_resistances table.

DustyBagel reviewed (commented):
Seems fine. This should probably be reviewed by someone who understands how this works better though.

DustyDave961 committed:
Update radiation.lua

Greater than or equal to 0 because negative makes no sense.

SmallJoker commented on technic/radiation.lua:
```diff
@@ -28,9 +28,26 @@ or complex internal structure should show no radiation resistance.
+    cache_radiation_resistance[node_name] = nil -- Invalidate cache
+end
+
+-- Function to register multiple node resistances at once
+function technic.register_multiple_resistances(resistances)
+    for node_name, resistance in pairs(resistances) do
+        technic.register_rad_resistance(node_name, resistance)
+    end
+end
```
Functions should be kept as generic as possible. This one can be very easily be implemented by modders. It does not bring any noticeable benefit.

DustyDave961 replied:
I suppose that's fair.

SmallJoker commented on technic/radiation.lua:
```diff
@@ -28,9 +28,26 @@ or complex internal structure should show no radiation resistance.
+local cache_radiation_resistance = {}
+
+-- Function to register node-specific resistance
+function technic.register_rad_resistance(node_name, resistance)
```
Please document this function in `doc/api.md`.

DustyDave961 replied:
I put it under the Helper functions. Will that work, or should it have its own section?

SmallJoker replied:
Fine by me.

SmallJoker commented on technic/radiation.lua:
```diff
@@ -165,38 +181,62 @@ local rad_resistance_node = {
+technic.register_group_resistance("wood", 1.7)
+
+-- Function to calculate radiation resistance
+function node_radiation_resistance(node_name)
```
This function should either be local, or in the `technic.` table and documented.

DustyDave961 replied:
I actually didn't change the name of that one since it's used to calculate radiation damage later on, but you do have a point.

SmallJoker commented on technic/radiation.lua:
```diff
@@ -165,38 +181,62 @@ local rad_resistance_node = {
+
+    -- Check for rad_resistance group in node definition
+    resistance = 0
+    for g, v in pairs(def.groups) do
+        if g == "rad_resistance" then
+            resistance = resistance + v
+        end
+    end
```
`technic.register_group_resistance` should be used instead of this special case.

DustyDave961 replied:
How would that even work? The whole point of the rad_resistance group is for individual nodes to have radiation resistance built into their node definitions. The group registration is for groups of nodes that share the same resistance value.

SmallJoker replied:
I was thinking of using the formula `resistance = rad_resistance_group[g] * v` where g is the group name and v the group value. The group value would multiply the base resistance.

That idea is however flawed. Group values can be used for other purposes, such as tool capabilities, thus have no relation to the radiance resistance.

Consider my original comment as invalid.

SmallJoker commented on technic_worldgen/nodes.lua:
```diff
@@ -106,7 +106,7 @@ minetest.register_node(":technic:lead_block", {
 	tiles = { "technic_lead_block.png" },
 	is_ground_content = true,
-	groups = {cracky=1, level=2},
+	groups = {cracky=1, level=2, rad_resistance=80},
```
`technic.register_rad_resistance` could be used here. Or register a group.

DustyDave961 replied:
Adding it to the node definitions seemed simpler for new nodes. I figured people could do it whatever way they wanted. Should I have stuck with just the registration functions?

SmallJoker replied:
This comment comes from the same underlying [flawed] thought as https://github.com/minetest-mods/technic/pull/651#discussion_r1801732331 .

This group is OK as-is.

DustyDave961 replied:
I suppose one reason I did it that way is because I haven't learned how to override groups without overwriting the original groups. Registering group resistances with the rad_resistance group sounds even more complicated.

SmallJoker replied:
It always depends on how much code you're saving vs the amount of care the code needs for maintenance and documentation. In this case I think it is OK to have a shorthand.

DustyDave961 committed:
Update api.md

The technic.register_rad_resistance function added

DustyDave961 committed:
Update radiation.lua

Unnecessary function removed.

DustyDave961 committed:
Update radiation.lua

Spaces bothered me.

DustyDave961 committed:
Update radiation.lua

The node_radiation_resistance function is only used locally.

DustyDave961 committed:
Update api.md

Documented rad_resistance group.

DustyDave961 committed:
Update radiation.lua

I guess we only need the rad_resistance group?

DustyDave961 committed:
Update radiation.lua

Space caused an error for some reason?

DustyDave961 committed:
Update api.md

Documented technic.register_group_resistance function

DustyDave961 committed:
Update nodes.lua

Added rad_resistance group to resistant nodes.

DustyDave961 committed:
Update radiation.lua

Technic specific node resistances moved to node definitions.

SmallJoker commented on technic/radiation.lua:
```diff
@@ -147,57 +158,49 @@ local rad_resistance_node = {
 	end
+end
+
+technic.register_group_resistance("concrete", 16)
+technic.register_group_resistance("tree", 3.4)
+technic.register_group_resistance("uranium_block", 500)
+technic.register_group_resistance("wood", 1.7)
```
Unfortunately this only works on nodes that are already registered when `technic` is loaded.
A few ideas to fix that:
1. re-add the previous `cache_radiation_resistance` to lazily calculate the node value when needed
2. calculate all values either in `core.register_on_mods_loaded` or in a `core.after(0, function() ??? end)`  callback (first server step)

DustyDave961 replied:
I found such functions in the Minetest lua api, but I'm unsure how to use the `minetest.register_on_mods_loaded(function())` function. Do I just replace "function" `with node_radiation_resistance`, or is there more to it?

SmallJoker replied:
@DustyDave961 [Changelog `0.4.16 → 5.0.0`](https://dev.minetest.net/Changelog) mentions this callback. Minetest 5.0.0 is already the minimal requirement for technic, thus this is not of a concern.

If functions are not mentioned in the changelog, you might as well have a look at the git blame of the `lua_api.md` line, which reveals when it was added.

SmallJoker commented on technic/radiation.lua:
```diff
@@ -28,9 +28,20 @@ or complex internal structure should show no radiation resistance.
 --]]
 
+-- Function to register node-specific resistance
+function technic.register_rad_resistance(node_name, resistance)
+	local node = minetest.registered_nodes[node_name]
+	if node then
+		if not node.groups then
+			node.groups = {}
+		end
+        node.groups.rad_resistance = resistance
+	end
+end
```
FYI: (no action needed in this case)
For this group it does not matter, but you'd generally want to use `core.override_item(node_name, { groups = newgroups })` so that the Minetest engine is aware of the groups as well. It is important for the client-side prediction of armor and tool digging groups (i.e. to determine how long it takes to dig a node with the selected tool).

DustyDave961 replied:
Does `minetest.override_item` remove groups that aren't included in the new groups list? I didn't use it because I was concerned about that.

SmallJoker replied:
`core.override_item` overrides the fields you're giving it, so here's how you can preserve them:

```Lua
local groups  = core.registered_nodes[node_name].groups or {}
groups.mygroupname = 123
core.override_item(node_name, groups)
```

DustyDave961 replied:
Thank you, that makes sense to me.

DustyDave961 committed:
Update radiation.lua

Why is whitespace such an issue? There's nothing there.

DustyDave961 committed:
Update radiation.lua

I thought I removed the white space already...

DustyDave961 committed:
Update radiation.lua

I think this will work?

DustyDave961 committed:
Update api.md

Radiation functions moved to Radiation section.

DustyDave961 committed:
Update radiation.lua

Shouldn't the resistance cache be initialized before the node_radiation_resistance function?

DustyDave961 committed:
Update nodes.lua

Changed "rad_resistance" to "rad_resistant"

DustyDave961 committed:
Update radiation.lua

Changed "rad_resistance" group to "rad_resistant".

DustyDave961 committed:
Update api.md

Changed rad_resistance group to rad_resistant. Why would anyone ever say "thumb of rule"?

DustyDave961 committed:
Merge pull request #1 from minetest-mods/master

Sync with Main Repository

DustyDave961 committed:
Update radiation.lua

Initialized the resistance cache inside the node_radiation_resistance function.
